airflow/providers/amazon/aws/triggers/base_trigger.py

```diff
@@ -0,0 +1,129 @@
+    :param status_queries: A list containing the JMESPath queries to retrieve status information from
+        the waiter response. See https://jmespath.org/tutorial.html
+
+    :param return_key: The key to use for the return_value in the TriggerEvent this emits on success.
+        Defaults to "value".
+    :param return_value: A value that'll be returned in the return_key field of the TriggerEvent.
+        Set to None if there is nothing to return.
+
+    :param waiter_delay: The amount of time in seconds to wait between attempts.
+    :param waiter_max_attempts: The maximum number of attempts to be made.
+    :param aws_conn_id: The Airflow connection used for AWS credentials. To be used to build the hook.
+    :param region_name: The AWS region where the resources to watch are. To be used to build the hook.
+    """
+
+    def __init__(
+        self,
+        *,
+        serialized_fields: dict[str, Any],
+        waiter_name: str,
+        waiter_args: dict[str, Any],
+        failure_message: str,
+        status_message: str,
+        status_queries: list[str],
+        return_key: str = "value",
+        return_value: Any,
+        waiter_delay: int,
+        waiter_max_attempts: int,
+        aws_conn_id: str | None,
+        region_name: str | None = None,
+    ):
+        # parameters that should be hardcoded in the child's implem
+        self.serialized_fields = serialized_fields
+
+        self.waiter_name = waiter_name
+        self.waiter_args = waiter_args
+        self.failure_message = failure_message
+        self.status_message = status_message
+        self.status_queries = status_queries
+
+        self.return_key = return_key
+        self.return_value = return_value
+
+        # parameters that should be passed directly from the child's parameters
+        self.waiter_delay = waiter_delay
+        self.attempts = waiter_max_attempts
+        self.aws_conn_id = aws_conn_id
+        self.region_name = region_name
+
+    def serialize(self) -> tuple[str, dict[str, Any]]:
+        # here we put together the "common" params,
+        # and whatever extras we got from the subclass in serialized_fields
+        params = dict(
+            {
+                "waiter_delay": self.waiter_delay,
+                "waiter_max_attempts": self.attempts,
+                "aws_conn_id": self.aws_conn_id,
+            },
+            **self.serialized_fields,
+        )
+        if self.region_name:
+            # if we serialize the None value from this, it breaks subclasses that don't have it in their ctor.
+            params["region_name"] = self.region_name
+        return (
+            # remember that self is an instance of the subclass here, not of this class.
+            self.__class__.__module__ + "." + self.__class__.__qualname__,
+            params,
+        )
+
+    @abstractmethod
+    def hook(self) -> AwsGenericHook:
+        """Override in subclasses to return the right hook."""
+        ...
```

Comment:
Not positive, but should this raise NotImplementedException like they did [here](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/sensors/sagemaker.py#L75) to enforce that?

Reply:
the `@abstractmethod` annotation is taking care of that (and more)

Reply:
oooohhhh.  I'll need to look into that more.    Sorry about that.